### ISSUE TITLE
Function attribute to disallow unwrap of returned value

### ISSUE THREAD

**Issue by camsteffen** [rust-lang/rust-clippy#17335]
### What it does

Sometimes you have a function where, if the caller unconditionally unwraps the returned value, there exists a better way to write that.

### Advantage

- Encourages consistency.
- Encourages better panic error messages since you can write it once and re-use rather than re-writing `expect("shouldn't happen") N times.

### Drawbacks

- People may be tempted to use the attribute for the wrong reason. It should not be used for "this is dangerous to unwrap" because that's rather subjective and that should be a decision left for the caller. Maybe there is a better attribute name?

### Example

```rust
#[clippy::no_unwrap(reason = "use `get_foo` instead")]
fn try_get_foo() -> Option<Foo> {
  ...
}

fn get_foo() -> Option<Foo> {
  try_get_foo().expect("really nice error message")
}

fn example() {
  let x = try_get_foo().unwrap(); // LINT
}
```

Could be written as:

```rust
fn example() {
  let x = get_foo();
}
```

### Additional Context

One tricky question: should it also lint on usages of `.expect("...")`? I think this would _usually_ be good, but I'm afraid it may cause some false positives where the caller intentionally uses a different error message. So I think I would not want it to lint `expect`.

Clippy should error/warn on using the attribute for a function that does not return Option or Result.
